File: puzzles/evernote/topFreqTerms/topFreqTermsEvernotePuzzle.cpp
#include "string"
#include "iostream"
#include "map"
#include "vector"

namespace std { using namespace __gnu_cxx; }

using namespace std;
int n, k; // I hate that

struct Tuple {
    string word;
    int occurences;

    Tuple() {
        word = "";
        occurences = -1;
    }

    Tuple(const Tuple& t) {
        word = t.word;
        occurences = t.occurences;
    }

    Tuple(string w, int o) {
        word = w;
        occurences = o;
    }
    
    bool operator >(Tuple other) {
#ifdef DBG
        cout << "> operator" << endl;
#endif
        return (occurences > other.occurences || (occurences == other.occurences && word < other.word));
    }

    bool operator <(Tuple other) {
#ifdef DBG
        cout << "< operator" << endl;
#endif
        return (occurences < occurences || (occurences == other.occurences && word > other.word));
    }
};

Tuple* EMPTY_TUPLE;

map<string, int> freq;
Tuple** kTop;

void try_insert(Tuple* item) {
    // item[0] = word
    // item[1] = occurences

    if (kTop[k] != EMPTY_TUPLE)
    {
        delete kTop[k];
    }
    kTop[k] = item; // Note : This index is not in the "top kTop", it's here because it's handy
    int i = k-1;
    while (i >= 0 && (*item) > (*kTop[i])) {
#ifdef DBG
        cout << "Swapping ! " << endl;
#endif
        // Swap values:
        Tuple* tmp = kTop[i];
        kTop[i] = kTop[i+1];
        kTop[i+1] = tmp;
        // Decrement:
        i -= 1;
    }
}

int main(int argc, char const *argv[])
{
// Grab number of lines:
    cin >> n;
// Count the words
    for (int x = 0; x < n; ++x)
    {
        string s;
        cin >> s;
        freq[s] = freq[s] + 1;
    }

    // Grab the top how many we are interested in
    cin >> k;

    kTop = new Tuple*[k+1];

    EMPTY_TUPLE = new Tuple();
    for (int i = 0; i < (k+1); ++i)
    {
        kTop[i] = EMPTY_TUPLE;
    }

    map<string, int>::iterator it;
    for (it = freq.begin(); it != freq.end(); ++it)
    {
        Tuple* item = new Tuple(it->first, it->second);
        try_insert(item);
    }

    for (int i = 0; i < k; ++i)
    {
        cout << kTop[i]->word << "\n";
    }

    delete EMPTY_TUPLE;

    return 0;
}

File: puzzles/evernote/topFreqTerms/topFreqTermsEvernotePuzzle.py
#!/usr/bin/python

# Grab number of lines:
n = int(raw_input())

# Count the words
freq = {}
for x in xrange(0, n):
    s = raw_input()
    if s in freq:
        freq[s] += 1
    else:
        freq.setdefault(s, 1)

# Grab the top how many we are interested in
k = int(raw_input())

# Solve the problem
kTop = [('A', float('-inf'))] * (k+1) # 'A' is the "smallest" letter in the alphabet 
# print kTop

def try_insert(item):
    global kTop, k

    # item[0] = word
    # item[1] = occurences

    kTop[k] = item # Note : This index is not in the "top kTop", it's here because it's handy
    i = k-1
    while (item[1] > kTop[i][1] or (item[1] == kTop[i][1]) and item[0] < kTop[i][0]) and i >= 0:
        # Swap values:
        tmp = kTop[i]
        kTop[i] = kTop[i+1]
        kTop[i+1] = tmp
        # Decrement:
        i -= 1

items = freq.items()
for x in items:
    try_insert(x)

for x in xrange(0, k):
    print kTop[x][0]